core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -394,30 +396,34 @@ private long checkDeletedAndGCed(VersionValue versionValue) {
     }
 
     @Override
-    public boolean index(Index index) {
-        final boolean created;
+    public void index(Index index) {
         try (ReleasableLock lock = readLock.acquire()) {
             ensureOpen();
             if (index.origin().isRecovery()) {
                 // Don't throttle recovery operations
-                created = innerIndex(index);
+                innerIndex(index);
             } else {
                 try (Releasable r = throttle.acquireThrottle()) {
-                    created = innerIndex(index);
+                    innerIndex(index);
                 }
             }
-        } catch (IllegalStateException | IOException e) {
+        } catch (IllegalStateException | IOException | VersionConflictEngineException e) {
             try {
-                maybeFailEngine("index", e);
+                if (maybeFailEngine("index", e) == false) {
+                    // we set operation specific engine failure that doesn't
```

Comment:
OK - I spent an hour researching this and this is the right thing to do, but we need way more explaination as to why :)

Here is a suggestion:

```
When indexing a document into Lucene, Lucene distinguishes between environment related errors (like out of disk space) and document specific errors (like analysis chain problems) by setting the IndexWriter.getTragicEvent() value for the former. maybeFailEngine checks for these kind of errors and returns true if that is the case. We use that information to indicate a document level failure and set the result on index.setFailure
```